### ISSUE TITLE
`holoviews.element. Curve` does not accept pandas data frame where index is `TimedeltaIndex`

### ISSUE THREAD

**Issue by miccoli** [holoviz/holoviews#2807]
The following code works as expected:

```
import pandas as pd
import holoviews as hv
hv.extension('bokeh', 'matplotlib')
import numpy as np

dt = pd.date_range('2018-02-23T12:34:45', '2018-02-23T12:47:00', freq='1s', name='t')
data = pd.DataFrame(np.sin((dt-dt[0]) / np.timedelta64(1,'m')).values, columns=('A', ), index=dt)

hv.Curve(data)
```
but 

```
data.index = dt - dt[0]
hv.Curve(data)
```

raises an error:

```
TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

I would like to plot pandas data frames indexed both by `DatetimeIndex` (works) or `TimedeltaIndex` (does not work)

**Comment by philippjfr** [comment]
Using the latest holoviews version this results in the following error in bokeh:

```python
~/development/bokeh/bokeh/core/properties.py in validate(self, value)
    695 
    696         if not (value is None or any(param.is_valid(value) for param in self.type_params)):
--> 697             raise ValueError("expected an element of either %s, got %r" % (nice_join(self.type_params), value))
    698 
    699     # TODO (bev) implement this

ValueError: expected an element of either Float, Datetime or Int, got Timedelta('0 days 00:00:00')
```

We'll either have to add timedelta support directly in bokeh or convert timedeltas before plotting.

**Comment by miccoli** [comment]
Thanks for the quick analysis.
Should I close this issue here and reopen on Bokeh?

**Comment by philippjfr** [comment]
>Should I close this issue here and reopen on Bokeh?

I think leaving it open here is sensible and it looks like there is an existing issue about this in bokeh (see https://github.com/bokeh/bokeh/issues/5426), so maybe just bump that issue since it **seems** straightforward to fix.

**Comment by miccoli** [comment]
Indeed a quick and dirty commit (see miccoli/bokeh@05457535) confirms that this issue can be solved by fixing bokeh/bokeh#5426 (a graph is produced without raising an exception). However the resulting graph is still not satisfactory for me (due to poor axis tick formatting), may be due to inconsistent `TimeDelta` support in bokeh, see also bokeh/bokeh#7867.

For now I will investigate this issue on the bokeh side, in order to see if it can be completely fixed there, and then I will come back here.

**Comment by philippjfr** [comment]
Looks like this is something we'll have to handle on our end for now, by ensuring timedelta types should not use a datetime axis.

**Comment by miccoli** [comment]
I've done some testing using my patched bokeh  miccoli/bokeh@0545753. Starting from numpy `datetime64` and `timedelta64` types, I've created two different pandas data frames.

```
dt = np.arange('2018-03-14T15:30:21', '2020-06-28T13:30:23', dtype='datetime64[h]')
td = dt - dt[len(dt)//2]
y = np.sin(4*np.pi*td / td.ptp())

pddt = pd.DataFrame(y, index=dt, columns=['sin'])
pddt.index.name = 'datetime'

pdtd = pd.DataFrame(y, index=td, columns=['sin'])
pdtd.index.name = 'timedelta'
```

Here the results of `hv.Curve(pddt)` and `hv.Curve(pdtd)`

![image](https://user-images.githubusercontent.com/6403904/41936978-51a59ba2-798f-11e8-8c6a-7d1bb44c85ff.png)

![image](https://user-images.githubusercontent.com/6403904/41937005-65c7687c-798f-11e8-8b9e-3d98f7e15626.png)

The problem here is that in bokeh timedelta are scaled as ms (millisecond) and represented with a plain linear axis. This is could be OK, but the choice of ms is not very sensible in my opinion... Using the matplotlib backend one gets the same linear axis but with timedelta scaled as ns (nanosecond).

![image](https://user-images.githubusercontent.com/6403904/41937693-4bca5c34-7991-11e8-926e-622bcf4d28c1.png)

Things are not that better in pandas: `pdtd.plot(figsize=(14,3), grid=True)`

![image](https://user-images.githubusercontent.com/6403904/41937835-b02c39f4-7991-11e8-8e81-a284a8605a73.png)

Things to fix, in my opinion:
1. consistent behaviour for bokeh and matplotlib
1. if a linear axis is to be used (without  dynamic scaling like the bokeh datetime axis), times should be given in seconds and properly labelled as so.

Of course the ideal solution would be to implement native timedelta for the datetime axis (or maybe create a new timedelta axis type.)

**Comment by philippjfr** [comment]
The formatting issues are subsumed by https://github.com/holoviz/holoviews/issues/3143 so I'll close this now.
